イベント設定ページの初期UIとデータ取得ロジックの実装（競技種別・種目・都道府県・対象年齢・イベント名・イベントURL・地域・性別）

toshinori-m opened the pull request:
### 概要
イベント設定ページの初期UIを構築し、必要なデータ（競技種別、種目、都道府県、対象年齢）を取得するロジックを実装しました。本PRは、2回に分けて行う実装の第1段階です。

### 変更内容
- EventSettingPage.tsx を新規作成
- 以下のフォーム項目をUIとして追加
  - 競技種別（単一選択）
  - 種目（複数選択）
  - 都道府県（単一選択）
  - 対象年齢（複数選択）
  - イベント名（テキスト入力）
  - イベントURL（URL入力）
  - 地域（テキストエリア）
  - 性別（ラジオボタン）
- エラー時のメッセージ表示を実装
- 文字数制限に応じた警告表示を実装（イベント名、地域）

close https://github.com/toshinori-m/stay_connect/issues/189

toshinori-m committed:
feat: イベント設定ページのUIおよび初期データ取得ロジックを実装

- 競技種別、種目、都道府県、対象年齢、イベント名、イベントURL、地域、性別に関するフォームUIおよびデータ取得処理を追加

hirosi1900day reviewed (commented):
SearchForm などでもselectを使っていたり他のところでinputなどを使っていると思いますが、デザインの一貫性やメンテナンス性を考えるとcomponents/SelectField.tsxやcomponents/InputField.tsxった風に汎用的に使えそうな小さなコンポーネントを作り、これを使い回すことでプロジェクト内のInputFieldで同じデザインに統一されてかつみやすいコードになるかなと思いましたがいかがでしょうか？？(もちろんpropsでclassなどは渡せて上書きや追加などをできるようにするイメージ)

割と各社こういった汎用的に使えるコンポーネントを作ってstorybookで管理しています。(今回storybookの導入までは不要ですが知識として参考までに)
https://zenn.dev/fullyou/articles/853b77a3ce9144

(汎用コンポーネントはあまり大きくせずbuttonならbuttonだけのコンポーネントを作るみたいなイメージ。もちろんボタンにもいろんな色があるのである程度の種類を切り替えれるようにするイメージ)

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingPage.tsx:
```diff
@@ -0,0 +1,277 @@
+    } catch {
+      setErrors(["競技・都道府県・対象年齢のデータ取得に失敗しました。時間を置いて再試行してください。"])
+    }
+  }
```
同じような処理がHomePage.tsxにもありますかね？？

toshinori-m replied:
はい、HomePage.tsxで同じ処理があったので 下の箇所でuseFetchInitialData として共通化していました。
- https://github.com/toshinori-m/stay_connect/pull/192#pullrequestreview-2708592906

ただ、その際に useActionState を使っていたため、「初期データ取得には適さないのでは？」というコメントもあり納得して削除しました。
- https://github.com/toshinori-m/stay_connect/pull/195#discussion_r2011961004

よろしくお願いします。

hirosi1900day replied:
そちらの指摘はuseActionStateはfrom actionの際に利用するものなので、useEffectでfetchするときに使用するような系統のものではないのでは？という指摘になります。
一方今回の処理はuseEffectで初期データ取得用のものなので共通化で良いと思いますがいかがでしょうか？

hirosi1900day replied:
以下のdocsのようにform action時に使うものがuseActionState
https://ja.react.dev/reference/react/useActionState#reference

toshinori-m replied:
分かりました。
ありがとうございます！

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingPage.tsx:
```diff
@@ -0,0 +1,277 @@
+                          ) || null
+                          setSelected(option)
+                        }
+                      }}
```
UIにはあまりロジックを入れたくない
やるなら適切な命名をした上で関数に切り出したりした方が良さそう

toshinori-m replied:
そうでした。申し訳ありません。

toshinori-m committed:
refactor: イベント設定フォームの構造改善

- useInitialFormData フックで初期データ取得処理を共通化
- useFetchDisciplines フックでスポーツ種目の取得処理を分離
- jsx内のUI ロジックを関数に切り出し

toshinori-m commented:
EventSettingPage.tsxと同じようにHomePage.tsxの修正を、下のissueで修正します。
・[HomePage.tsxのスタイル統一（初期データ取得用の共通化、jsx内のロジックを無くすなど）](https://github.com/toshinori-m/stay_connect/issues/198)

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingPage.tsx:
```diff
@@ -0,0 +1,232 @@
+    <div className="flex items-center justify-center mt-32 md:mt-20">
+      <div className="w-full md:w-3/5 xl:w-2/5 shadow-gray-200 bg-sky-100 rounded-lg">
+        <h2 className="text-center mb-10 pt-10 font-bold text-3xl text-blue-600">イベント設定</h2>
+        <form className="px-4 md:px-0 text-center">
```
formを使っているのにも関わらずsubumitボタンがないのは想定通りの認識であってますか？？

toshinori-m replied:
はい、合っています
formが無いと下のように感じるかと思い記しておきました。
・このjsxって表示だけするもの？
・このEventSettingPage.tsxtって何するもの？

よろしくおねがいします

hirosi1900day reviewed (approved):
多分次のPRでformの送信処理を実装するのであろうと予測してapprove
PRを分ける際は次のPRでどう言う処理を書くつもりかを書いていただけるとわかりやすいかもです(自分が見逃しているだけだったらすみません)

toshinori-m commented:
> イベント設定ページの初期UIを構築し、必要なデータ（競技種別、種目、都道府県、対象年齢）を取得するロジックを実装しました。本PRは、2回に分けて行う実装の第1段階です。

@hirosi1900day 
本PRの概要に上のように書いておけば分かりやすいかと思ったのですが、次のPRで下の追加を行う予定です。

「開始日付」「終了日付」「募集チーム数」「イベント目的」「その他」これらの情報を送信するボタン

なお、次からPRを分けるときは追加で行う作業も記しておきます

よろしくお願いします
